src/Runtime/Core/Serialization/InferDataContractResolver.cs

```diff
@@ -39,11 +154,428 @@ public override Type ResolveName(string typeName, string typeNamespace, Type dec
+        {
+            // Type lists are delimited with square brackets, as is each type in the list, and they are separated by commas.
+            // For example, X`4[[A],[B],[C],[D]]
+            var list = new List<TypeId>();
+
+            if (typeString.Length == cursor)
+            {
+                return (cursor, new TypeId[0]);
+            }
+
+            // We check for the marker that is used for type lists and arrays.
+            if (typeString[cursor] != '[')
+            {
+                return (cursor, new TypeId[0]); 
+            }
+
+            // We verify that this is not a marker of a single dimensional array.
+            if (typeString[cursor + 1] == ']')
+            {
+                return (cursor, new TypeId[0]);
+            }
+
+            // We verify that this is not a marker of a multi-dimensional array.
+            // (After this, it must be a type list marker.)
+            if (typeString[cursor + 1] == ',')
+            {
+                return (cursor, new TypeId[0]);
+            }
+
+            cursor++;
+
+            while (true)
+            {
+                if (typeString[cursor] != '[')
+                {
+                    throw new InvalidOperationException("invalid type list element.");
+                }
+                cursor++;
+
+                TypeId item;
+                (cursor, item) = ReadTypeName(typeString, cursor);
+                list.Add(item);
+
+                if (typeString.Length < cursor)
+                {
+                    throw new InvalidOperationException("invalid cursor.");
+                }
+
+                if (typeString.Length == cursor)
+                {
+                    throw new InvalidOperationException("truncated type string.");
+                }
+
+                // Check for the closing bracket of the type string.
+                if (typeString[cursor] != ']')
+                {
+                    throw new InvalidOperationException("invalid type string.");
+                }
+
+                cursor++;
+
+                // there is not another item.
+                if (typeString[cursor] != ',')
+                {
+                    break;
+                }
+
+                cursor++;
+            }
+
+            if (typeString[cursor] != ']')
+            {
+                throw new InvalidOperationException("invalid type list termination.");
+            }
+            cursor++;
+
+            return (cursor, list.ToArray());
+        }
+
+        /// <summary>
+        /// Parse a type string into a type layout which gives basic information
+        /// about the makeup of the type.
+        /// </summary>
+        /// <param name="typeString">The type string.</param>
+        /// <returns>
+        /// Basic information about how the type is constructed from other types.
+        /// </returns>
+        internal static TypeId GetTypeLayout(string typeString)
```

Comment:
If you are returning a `TypeId` then it should be called `GetTypeId`

Reply:
Changed to ParseTypeString to be consistent with change of name from TypeId to ParsedTypeString.